ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +64,221 @@ data NtpStatus =
+        Just offset -> do
+            traceWith tracer $ NtpTraceUpdateStatusClockOffset $ getNtpOffset offset
+            atomically $ writeTVar ntpStatus $ NtpDrift offset
+    traceWith tracer NtpTraceClientSleeping
+    threadDelayInterruptible ntpStatus $ fromIntegral $ ntpPollDelay ntpSettings
     where
-    closeSockets :: Sockets -> IO ()
-    closeSockets sockets = do
-        case sockets of
-            This a -> fn a
-            That a -> fn a
-            These a b -> fn a >> fn b
-        traceWith tracer NtpTraceSpawnNtpClientSocketsClosed
-
-    fn :: Last (WithAddrFamily t Socket.Socket) -> IO ()
-    fn (Last sock) = Socket.close $ runWithAddrFamily sock
-
-    -- Try to resolve addresses, on failure wait 30s and start again.
-    resolve = do
-        traceWith tracer NtpTraceSpawnNtpClientResolveDNS
-        addrs <- catMaybes <$> traverse (resolveNtpHost tracer) (ntpServers settings)
-        if null addrs
-          then do
-            atomically $ writeTVar ncStatus NtpSyncUnavailable
-            -- either wait 30s or wait for `NtpSyncPending` which might be set
-            -- by a client (e.g. wallet), in which case try to resolve the dns.
-            race_
-              (threadDelay 30000000)
-              (do
-                atomically $ do
-                  s <- readTVar ncStatus
-                  case s of
-                    NtpSyncPending -> return ()
-                    _              -> retry
-                traceWith tracer NtpTraceSpawnNtpClientResolvePending
-              )
-            resolve
-          else return addrs
-
--- Try to create IPv4 and IPv6 socket.
-mkSockets :: Tracer IO NtpTrace -> NtpClientSettings -> IO Sockets
-mkSockets tracer settings =
-    doMkSockets `catch` handleIOException >>= \case
-        Just socks -> pure socks
-        Nothing     -> do
-            traceWith tracer NtpTraceMkSocketsNoSockets
---            logWarning "Couldn't create either IPv4 or IPv6 socket, retrying in 5 sec..."
-            threadDelay 5000000
-            mkSockets tracer settings
+        runProtocol _version _localAddr [] = return []
+        runProtocol _version Nothing    _  = return []
+        runProtocol version (Just addr) servers = do
+            socketAction tracer ntpSettings addr servers >>= \case
+                Left err -> do
+                    traceWith tracer $ NtpTraceRunProtocolError version err
+                    return []
+                Right [] -> do
+                    traceWith tracer $ NtpTraceRunProtocolNoResult version
+                    return []
+                Right r@(_:_) -> do
+                    traceWith tracer $ NtpTraceRunProtocolSuccess version
+                    return r
+
+socketAction ::
+       Tracer IO NtpTrace
+    -> NtpClientSettings
+    -> AddrInfo
+    -> [AddrInfo]
+    -> IO (Either IOError [ReceivedPacket])
+socketAction tracer netSettings localAddr destAddrs
```

Comment:
Maybe there's a more descriptive name for this function. What about `requestNtpMeasurements`?

It would be also good to add haddoc documentation what it does in a descriptive manner.

Reply:
`queryServers` ?

Reply:
or `runNtpQueries`